### ISSUE TITLE
CmsSigner.ComputeSignature() creates a signature with incorrect SignerInfo

### ISSUE THREAD

**Issue by alexcyoung** [dotnet/corefx#29691]
I'm currently porting over some code to .Net Core and using the recently added CmsSigner from System.Security.Cryptography.Pkcs (I'm using the 4.5.0-rc1)  The certificate I'm using has a serial number of FD319CB1514B06AF49E00522277E43C8, it also happens to be self signed cert. 

Here is what I'm doing to create the signature. 
```
X509Certificate2 cert = new X509Certificate2(certificate, password);
ContentInfo contentInfo = new ContentInfo(message);
SignedCms signedCms = new SignedCms(contentInfo, true); 
CmsSigner cmsSigner = new CmsSigner(cert);
cmsSigner.SignerIdentifierType = SubjectIdentifierType.IssuerAndSerialNumber;

signedCms.ComputeSignature(cmsSigner);
signature = signedCms.Encode();
```

After calling ComputeSignature the signedCms object created has the wrong serial number in the SignerInfos list, with "00" prefix being added. 
Calling 
`signedCms.SignerInfos[0].SignerIdentifier.Value`

Produces the result below, where you can see the extra "00" being added
{System.Security.Cryptography.Xml.X509IssuerSerial}
    IssuerName: "CN=myorg"
    SerialNumber: "**00**FD319CB1514B06AF49E00522277E43C8"

In addition, if I check the signedCms.Certificates[0].SerialNumber object I can see the correct serial number with no "00" prefix. 

Can you tell me what is causing this "00" to be added and if its a bug or something I can correct with config.

**Comment by bartonjs** [comment]
Correction, it _is_ an encoding problem.  Apparently I was looking at the encoded serial number of the certificate (unsurprisingly: had the value of the encoded serial number of the certificate) instead of the IssuerAndSerialNumber value, which does have the leading 0x00.

**Comment by alexcyoung** [comment]
I wasn't aware the serial number was invalid although it was created by makecert. 

I should add that the same certificate and code combination works in 4.5 without issue. 

Appreciate you looking into this.

**Comment by alexcyoung** [comment]
Hi @bartonjs so looking at the fix, it's got the 2.2.0 milestone. Does this mean there's not chance of it making the 2.1 release? If so can you recommend a workaround I can use for this?

**Comment by bartonjs** [comment]
That PR would be marked as 2.2 anyways, since it's a PR to master.  That said, I was told that to make it into 2.1 at this point a bug pretty nearly has to break Hello World, so your assessment is correct anyways :).

There are a couple of "workarounds" that I can think of:
* Generate a new certificate whose serial number does not have the high bit set.
  * That's not always easy, but maybe you could look at it as practice for handling private key compromise? :)
* Since System.Security.Cryptography.Pkcs is a standalone package (not part of the shared framework) you could reference a version from the 2.2/master feed on myget.org after the fix makes it in.
  * There's some risk to that, since there's a pretty good probability that the package will start calling into API new in 2.2 that is part of the shared framework.  And if there were any security updates to the package you wouldn't see them
* Make a private build of 2.1 + this fix
  * Doesn't have the "at some point a newer package starts getting MissingMethodException" problem that pulling from 2.2 does, but does have the "you're detached from the security update process" problem.
* Wait until 2.2, or until I can get the fix into a 2.1 quality update.
* Investigate if Bouncy Castle, or some other crypto suite, can tide you over

**Comment by alexcyoung** [comment]
Understand on not making it into 2.1, a line needs to be drawn somewhere otherwise things would never get released :). 

The first option sounds like it might be the best, as you say, good practice for replacing a compromised key. Will also have a look at BC too to see if its an option. 

Out of interest, what would the likely how/ time period of this making into a 2.1 quality update? (I appreciate that release schedules are probably not your area, just asking a best guess). 

Really appreciate your help and responsiveness on this .

**Comment by danmosemsft** [comment]
@alexcyoung can you please confirm you were able to get unblocked using one of the workarounds? That is part of what we have to consider when deciding whether to backport this fix into 2.1.

**Comment by danmosemsft** [comment]
@joshfree also pointed out that makecert is documented to be for testing purposes only. https://msdn.microsoft.com/en-us/library/bfsktky3(vs.80).aspx  Nevertheless, if it is prone to producing certs we can't consume, that is still a problem.

**Comment by alexcyoung** [comment]
The only option we've looked at is replacing the certificate with one that did not use the high bit set at the start. This has been successful during testing.

I understand from earlier comments that usage of the high bit set falls outside the definition in the RFC for certificate serial numbers. However this certificate was generated by makecert (and makecert designated the serial number, there was no special config used during creation) so I would suggest that if the issue is not corrected in 2.1 the situation I've encountered would become fairly common.

**Comment by danmosemsft** [comment]
@alexcyoung thanks. Is it possible to help me answer: does makecert always make certs with this issue, or if not how often do you see it?

**Comment by danmosemsft** [comment]
Decision was made to defer this from 2.1 servicing unless more evidence emerges that it can't easily be worked around or is otherwise more impactful than we thought. It would be good to know how often makecert does this.

**Comment by alexcyoung** [comment]
Apologies for the delay I've been away from my computer until today. I can look at this tomorrow and get you some indication of how common the issue is.

**Comment by alexcyoung** [comment]
I've carried out a series of 20 makecert executions allowing the tool to designate the serial number. If I'm understanding things correctly it looks like the high bit set is used in 15 of the 20 serial numbers so 75% of the time. I've included the numbers below. 

I don't argue that the work around is fairly simple (you can designate a serial number in make cert yourself) but actually finding the issue is not simple and certainly not quick (it took me over two days). This is mainly due to the fact that when using CmsSigner the issue is not apparent and the operation goes ahead without a hitch. You only become aware of the issue when you come to check the signature and even then the actual information is pretty buried and not something a developer would normally need to be aware of. I only found it because I used a 3rd party tool that flagged it couldn't find the serial number. Had I just been debugging I'm not sure I would have ever found it. 

So in short, I would argue that for many developers the symptoms of the problem may not always be associated with the actual cause it will just appear like "it doesn't work". I also understand what you say about makecert only being used for testing and development. However, that's exactly what most developers do and hence will use makecert initially with an official certificate only being used once a piece of software is production ready. 

On this basis I would ask you to reconsider your decision and backport this fix into 2.1

98fb53d041a8f4b64b042909f099d828
ed455da7d4f5bbba4d07a749f51f9742
c3dbb0ce458621b049a0453c7a179024
954c1cf7703a898d4d165169b00d50d2
816cda9333e71baa4b05766b1b955ac1
a4eea2f6988ea8a943b42cd3f2ac0ff0
ef20ff946f697f8c46543cdc9d5ab96c
cddcb737e1e786a64ac7b36042a427d0
c3e2b00958bd738e447fe13a4ad5eaff
1ef09594cb8584874fe412b6cae29e84
cc7bf57c646e8aa64f8b24a3f27fa72d
f81facfff08d228446e2ef5fe6e1ddfc
5b5ded70e3fa2a8c40a9b05a285586d4
ed3a3ffd27780a8d47826f0939708918
a0647b97ebc50fa4480f91dff0b0b50d
eecfb4775c4eebbb4bd4b69769d835e1
89f137fdbb530daf468379669b027e42
38d8b8e9a8ef61a04738ae347810b4e1
31f0ccaffdd4509946344127437c5bb3
e941b4d3a9fefd8d42ccbb1314887ea3

**Comment by bartonjs** [comment]
@danmosemsft 

> It would be good to know how often makecert does this.

It looks like makecert calls `CoCreateGuid` if no explicit serial number was provided.  That then is interpreted by Windows to be a little-endian big integer.  The bit in question (which is the high bit of the last byte in the guid (either the blitted form or the string-hex form)) is random, so the amortized statement is: `50% of certificates generated by makecert.exe will fail to interoperate`.

**Comment by danmosemsft** [comment]
Thanks @alexcyoung @bartonjs .  I've marked it to consider for the next servicing (it will miss 2.1.1 I think)

@bartonjs when you return, perhaps you could open a bug against makecert owners? Looks like it's in ds/security/.../makecert.cpp in Windows. I do see the CoCreateGuid in there. I was going to open one, but I figured you have more street cred over there.

**Comment by karelz** [comment]
Fixed in 2.1 servicing in PR #30381. Closing.
It will be part of 2.1.3 release.

**Comment by danmosemsft** [comment]
@karelz actually this went into 2.1 without approval (due to miscommunication). @bartonjs is going to talk to shiproom tomorrow

**Comment by karelz** [comment]
OK, I just reflected the status of the branch :) ... if we back it out, we can reopen and adjust milestone.
